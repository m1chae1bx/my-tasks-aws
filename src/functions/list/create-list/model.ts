import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateListRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    name: string;
    isDefault?: boolean;
  };
}

export const createListRequestSchema: JSONSchemaType<CreateListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 100,
        },
        isDefault: {
          type: "boolean",
          nullable: true,
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createListRequestSchema);
