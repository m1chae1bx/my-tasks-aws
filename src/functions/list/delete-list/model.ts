import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteListRequest {
  pathParameters: {
    userId: string;
    listId: string;
  };
}

export const deleteListRequestSchema: JSONSchemaType<DeleteListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
        listId: {
          type: "string",
        },
      },
      required: ["userId", "listId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteListRequestSchema);
