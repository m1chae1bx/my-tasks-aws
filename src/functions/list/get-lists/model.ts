import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetListsRequest {
  pathParameters: {
    userId: string;
  };
}

export const getListsRequestSchema: JSONSchemaType<GetListsRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getListsRequestSchema);
