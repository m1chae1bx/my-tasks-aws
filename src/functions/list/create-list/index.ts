import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : undefined,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const userId = request.pathParameters.userId;
    const { name, isDefault } = request.body;

    const list = new List(name, userId, isDefault);
    const id = await list.save();
    const response = { message: `List ${name} was created successfully`, id };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while creating the list. Please try again later.`
    );
  }
};
