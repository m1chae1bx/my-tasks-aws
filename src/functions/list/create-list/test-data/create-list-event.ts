import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const createListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
  body: JSON.stringify({
    name: "testListName",
    isDefault: false,
  }),
};
