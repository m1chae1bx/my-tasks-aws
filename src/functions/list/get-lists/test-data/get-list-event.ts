import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const getListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
};
