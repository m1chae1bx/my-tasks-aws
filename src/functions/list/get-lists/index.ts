import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad Request",
          errors: validateRequest.errors,
        }),
      };
    }
    const userId = request.pathParameters.userId;

    const response = await List.getAll(userId);
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while getting the task lists. Please try again later.`
    );
  }
};
