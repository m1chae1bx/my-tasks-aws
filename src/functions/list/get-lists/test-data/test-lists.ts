import { List } from "/opt/nodejs/list.model";

export const testLists: Partial<List>[] = [
  {
    id: "dfab1ffd-d3f3-4c45-b39a-46fcfd32fe7d",
    name: "List 1",
    userId: "testUserId",
    isDefault: false,
  },
  {
    id: "7a7a528c-142f-4b77-a168-8d6b8b56e40e",
    name: "List 2",
    userId: "testUserId",
    isDefault: false,
  },
];
