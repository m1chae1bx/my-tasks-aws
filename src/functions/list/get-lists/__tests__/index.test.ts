import { ajvError } from "@test-utils/base-objects";
import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "..";
import { validateRequest } from "../model";
import { getListEvent } from "../test-data/get-list-event";
import { testLists } from "../test-data/test-lists";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("/opt/nodejs/list.model", () => ({
  List: {
    getAll: jest.fn(),
  },
}));

jest.mock("/opt/nodejs/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("get-list", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (List.getAll as jest.Mock).mockResolvedValueOnce([testLists]);
      result = await handler(getListEvent);
    });

    it("should return a 200 response", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return the list", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain(testLists[0].id);
      expect(result.body).toContain(testLists[1].id);
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(getListEvent);
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the error", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - error getting the lists", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (List.getAll as jest.Mock).mockRejectedValueOnce(new Error("error"));
      (genericErrorHandler as unknown as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(getListEvent);
    });

    it("should return a 500 response", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return the error", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
