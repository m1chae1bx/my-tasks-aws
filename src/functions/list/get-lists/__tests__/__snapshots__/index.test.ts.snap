// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-list happy path should return the list 1`] = `"[[{\\"id\\":\\"dfab1ffd-d3f3-4c45-b39a-46fcfd32fe7d\\",\\"name\\":\\"List 1\\",\\"userId\\":\\"testUserId\\",\\"isDefault\\":false},{\\"id\\":\\"7a7a528c-142f-4b77-a168-8d6b8b56e40e\\",\\"name\\":\\"List 2\\",\\"userId\\":\\"testUserId\\",\\"isDefault\\":false}]]"`;

exports[`get-list sad path - error getting the lists should return the error 1`] = `"test error"`;

exports[`get-list sad path - invalid request should return the error 1`] = `"{\\"message\\":\\"Bad Request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;
